torch/nested/_internal/offload_tensor.py

```diff
@@ -0,0 +1,225 @@
+        )
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        # Doing any operation on a CachedTensor automatically calls restore and then unwraps
+        kwargs = {} if kwargs is None else kwargs
+        unwrapped_args = pytree.tree_map_only(
+            OffloadTensor, lambda x: x.restore(), args
+        )
+        unwrapped_kwargs = pytree.tree_map_only(
+            OffloadTensor, lambda x: x.restore(), kwargs
+        )
+        out = func(*unwrapped_args, **unwrapped_kwargs)
+
+        return out
+
+torch.serialization.add_safe_globals([OffloadTensor])
+
+
+_global_is_pending_offload = False
+
+
+def request_offload_all():
+    # Before the next creation of a new offload tensor, we want to offload all
+    # existing offload tensors.
+    # If you request offload all in the compiled region, the offloading will
+    # happen after compiled region is done.
+    global _global_is_pending_offload
+    _global_is_pending_offload = True
+
+
+class OffloadTensorRegistry:
+    def __init__(self, offload_hook: Optional[Callable] = None):
+        self.wrapper_refs = []
+        # Upon execution, the hook
+        # hook(host_tensor, device_tensor) -> None
+        self.offload_hook = None
+
+    def create(self, device, device_tensor, host_tensor):
+        ret = OffloadTensor(
+            device=device,
+            device_tensor=device_tensor,
+            host_tensor=host_tensor,
+            offload_hook=self.offload_hook,
+        )
+        self.wrapper_refs.append(weakref.ref(ret))
+        return ret
+
+    def maybe_offload_all(self):
+        global _global_is_pending_offload
+
+        if _global_is_pending_offload:
+            still_alive = []
+            for ref in self.wrapper_refs:
+                if (t := ref()) is not None:
+                    t.offload()
+                    still_alive.append(weakref.ref(t))
+            self.wrapper_refs = still_alive
+            _global_is_pending_offload = False
+        return
+
+    def copy(self):
+        # Is this what we want to do?
+        ret = OffloadTensorRegistry()
+        ret.wrapper_refs = self.wrapper_refs.copy()
+
+
+_global_offload_tensor_registry = None
+
+
+def init_offload_tensor_registry(registry):
+    global _global_offload_tensor_registry
+    _global_offload_tensor_registry = registry
+
+
+@contextmanager
+def offload_tensor_registry(registry):
+    global _global_offload_tensor_registry
+    # Save the current registry to restore it later
+    old_registry = _global_offload_tensor_registry
+    # Set the new registry
+    _global_offload_tensor_registry = registry
+    try:
+        yield
+    finally:
+        # Restore the old registry
+        _global_offload_tensor_registry = old_registry
+
+
+def create_offload_tensor(device, device_tensor, host_tensor):
+    assert _global_offload_tensor_registry is not None
+    if _try_get_fake_mode(device_tensor if device_tensor is not None else host_tensor):
```

Comment:
what does _try_get_fake_mode check? Is it to check if the tensor was used in a compile context?